File: backend/domains/map/services/data/climateData.js
// ============================================================================
// Climate sample helpers
// ----------------------------------------------------------------------------
// Single source of truth for reading the hourly climate array stored on a
// trip_day. The raw samples arrive nested ({ time, climate: { climate: {...} } })
// depending on where they were produced, so every consumer must unwrap them the
// same way. Anything that needs "how was the weather that day" in numbers
// belongs here; anything that turns numbers into prose belongs in
// naturalLanguage/.
// ============================================================================

/** Pull the inner weather record from a climate sample (handles the nesting). */
export function innerClimate(sample) {
  if (!sample) return null;
  const c = sample.climate || sample;
  return c.climate || c;
}

/** Unwrap a whole climate array into plain weather records, dropping blanks. */
export function climateRecords(climateArray) {
  if (!Array.isArray(climateArray)) return [];
  return climateArray.map((s) => innerClimate(s)).filter(Boolean);
}

/** Unwrap a climate array keeping the sample timestamp alongside the record. */
export function timedClimateRecords(climateArray) {
  if (!Array.isArray(climateArray)) return [];
  return climateArray
    .map((s) => ({ time: s.time, weather: innerClimate(s) }))
    .filter((s) => s.weather);
}

/** Arithmetic mean of the finite numbers in the list, or null when there are none. */
export function meanOf(nums) {
  const xs = (nums || []).filter((n) => typeof n === 'number' && Number.isFinite(n));
  if (xs.length === 0) return null;
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

/** Sum of the finite numbers in the list (0 when empty). */
export function sumOf(nums) {
  return (nums || [])
    .filter((n) => typeof n === 'number' && Number.isFinite(n))
    .reduce((a, b) => a + b, 0);
}

/**
 * Day-level weather aggregates used by the state engine and the prompt blocks.
 * @param {Array} climateArray - hourly samples as stored on trip_days.climate
 * @returns {{ meanTemperature: number|null, meanWind: number|null,
 *   meanCloud: number|null, totalPrecipitation: number }}
 */
export function climateStats(climateArray) {
  const records = climateRecords(climateArray);
  return {
    meanTemperature: meanOf(records.map((w) => w.temperature_2m)),
    meanWind: meanOf(records.map((w) => w.wind_speed_10m)),
    meanCloud: meanOf(records.map((w) => w.cloud_cover)),
    totalPrecipitation: sumOf(records.map((w) => w.precipitation || 0)),
  };
}

File: story-engine/app/climate_data.py
# ============================================================================
# Climate sample helpers
# ----------------------------------------------------------------------------
# Port of backend/domains/map/services/data/climateData.js (pure functions,
# no DB — reused here directly instead of going through an HTTP adapter,
# since they operate only on the `climate` array already embedded in the
# `day` payload).
# ============================================================================


def inner_climate(sample):
    """Pull the inner weather record from a climate sample (handles the nesting)."""
    if not sample:
        return None
    c = sample.get('climate') or sample
    return c.get('climate') or c


def climate_records(climate_array):
    """Unwrap a whole climate array into plain weather records, dropping blanks."""
    if not isinstance(climate_array, list):
        return []
    return [r for r in (inner_climate(s) for s in climate_array) if r]


def timed_climate_records(climate_array):
    """Unwrap a climate array keeping the sample timestamp alongside the record."""
    if not isinstance(climate_array, list):
        return []
    result = []
    for s in climate_array:
        weather = inner_climate(s)
        if weather:
            result.append({'time': s.get('time'), 'weather': weather})
    return result


def mean_of(nums):
    """Arithmetic mean of the finite numbers in the list, or None when there are none."""
    xs = [n for n in (nums or []) if isinstance(n, (int, float))]
    if len(xs) == 0:
        return None
    return sum(xs) / len(xs)


def sum_of(nums):
    """Sum of the finite numbers in the list (0 when empty)."""
    return sum(n for n in (nums or []) if isinstance(n, (int, float)))
